### DISCUSSION TITLE
Unable to get sprite to animate

### DISCUSSION THREAD

**Discussion by Vivraan** [pixijs/pixijs#7144] [Q&A]
### Migrated from
https://github.com/pixijs/pixi.js/issues/7136

### Expected Behavior
Player walking animation as in:

https://user-images.githubusercontent.com/12777564/104153882-7fc52f00-5409-11eb-83fd-a76aa05a7845.mp4

### Current Behavior
Player switching animations but not changing frames in:

https://user-images.githubusercontent.com/12777564/104153966-aaaf8300-5409-11eb-8e3f-d6ca85736b4e.mp4

Often, if I change the frame using `gotoAndPlay()`, the frame is set to this index in spite of emptying the cache while refreshing.

https://user-images.githubusercontent.com/12777564/104154122-1eea2680-540a-11eb-8df4-257ef5902bbe.mp4

My code:
https://gitlab.com/Vivraan/mario50-pixijs/-/blob/227a18672242d9c468274e6a50c53cff1adda7f2/player.js#L63-73
```js
        const move = (dt) => {
            if (this._vx != 0) {
                this._anim.textures = animations.walking
                if (!this._anim.playing) {
                    this._anim.play()
                }
                this._anim.scale.x = Math.sign(this._vx)
            } else {
                this._anim.textures = animations.idle
            }
        }
```

### Steps to Reproduce
The commit I'm working from is [this one](https://gitlab.com/Vivraan/mario50-pixijs/-/commit/227a18672242d9c468274e6a50c53cff1adda7f2), which summarises the repro steps.

### Replies from Issues thread
@bigtimebuddy says:
I think your logic for move incorrectly calls every frame. This mean that it's constantly setting textures on your animation frame after frame. If you cache the previous direction to only handle changes, it's better.

Here's an isolated example (click on the character on the right and then use the left and right arrow keys to see the animation switch). https://jsfiddle.net/bigtimebuddy/qo12kdnj/

This is not a limitation of the AnimatedSprite. In your case, this is an appropriate use.

**Comment by Vivraan** [comment]
@bigtimebuddy I tried to take your code (I got to learn just from viewing it) but I still can't get it to work.

**Reply by Vivraan** [reply]
Here's how I'm using the components from your fiddle:
([Commit on GitLab](https://gitlab.com/Vivraan/mario50-pixijs/-/commit/b790f27ce867c5277642a6531a825ba90e153593))
```js
// From https://jsfiddle.net/bigtimebuddy/qo12kdnj/
        let isLeft = false
        let isRight = false

        const onKeypress = ({ type, key }) => {
            const isKeyDown = type === 'keydown'
            switch (key) {
                case INPUT_MOVE_LEFT:
                    isLeft = isKeyDown
                    break
                case INPUT_MOVE_RIGHT:
                    isRight = isKeyDown
                    break
            }
        }

        window.addEventListener('keydown', onKeypress)
        window.addEventListener('keyup', onKeypress)

        let lastDirection = -1

        const move = () => {
            let direction = Number(isRight) - Number(isLeft)

            if (direction !== lastDirection) {
                if (direction === 0) {
                    this._animState = 'IDLE'
                    this._anim.textures = animations.idle
                    this._vx = 0
                    this._anim.scale.x = this._anim.scale.y
                } else {
                    this._animState = 'WALKING'
                    this._anim.textures = animations.walking
                    this._vx = MOVE_SPEED * direction
                    this._anim.scale.x = this._anim.scale.y * direction
                }
                this._anim.play()
                lastDirection = direction
            }
        }

        this._behaviour = Object.freeze({
            IDLE: move,
            WALKING: move,
            IN_AIR: () => { }
        })

        this._app.ticker.add(() => {
            const dt = deltatime(this._app.ticker)
            this._behaviour[this._animState].call()
            this._anim.update(dt)
...
```

**Reply by Vivraan** [reply]
Now my entire movement isn't working, even after I began updating `this._vx`!

**Comment by Vivraan** [comment]
Anyone? @bigtimebuddy?

Apologies for tag-spamming in so many places!

**Reply by NikolasP98** [reply]
I'm having the same issue... The animated sprite is moving, but frames are stuck on zero and aren't looping. My code only changes animations when needed, or else it escapes the re-assignment of textures. Directions change as well. Here's my code: https://pastebin.com/9kzu7bmJ and a live example: https://portfolio-e-five.vercel.app/peeps
